### ISSUE TITLE
Create funtional tests in relation to #issue 868

### ISSUE THREAD

**Issue by drunkel** [MarkUsProject/Markus#1199]
There should be some tests that explicitly test that grading the same submission multiple time does not lead to a repeat of issue 868.

**Comment by MatakuT** [comment]
I was not successful at implementing the proper test process for this issue. I will post what I have tried here, so hopefully the next person will have a better idea.
The basic principle if quite simple. Two tests have to be made for 'collect_all_submissions' and 'collect_ta_submissions' inside of submission_controller.rb
Let's talk about one: 'collect_all_submissions' as it is simpler of the two.
The first call to collect has to happen and then 'graded_submission_results' is called to pick up the counter. The second collect with the same set up should not change the counter. So after the calls two counters are compared and they have to be equal.

The object is set via Rspec. This object has an instance of SubmissionCollection with two grouping queues: priority_queue and results_queue. Those grouping queues contain at least 1 grouping (2 in assignment_spec.rb).
When the object is passed to 'collect_all_submissions' it is then goes to 'push_groupings_to_queue(groupings)' in submission_collector.rb  which initializes local grouping queues to the ones passed into the function. Then in the last bit of 'push_groupings_to_queue()' start_collection_process is called which forks a child process. This child process either does not properly copy the information from its parent or gets corrupt somewhere along the way. The child process calls collect_next_submission() which crushes as it's not able to access the grouping queues that were set up previously. 
ERROR: /home/vagrant/Markus/app/models/submission_collector.rb:47:in `priority_queue': undefined method `groupings' for nil:NilClass (NoMethodError)
I just want to empahsize again that this error happens in the child, not the initial call to push_groupings_to_queue()
It seems to me as Rspec and multi-threading do not get along very well. 
There is a bit in 'start_collection_process' that tells if the platform is Windows then the forking should be skipped. My idea is that skipping forking for testing should solve the problem.

I'm attaching the two tests that I created. They are the same just different design decisions. One test is called from controller_spec and another from assignment model_spec. 
![submissions_controller_test](https://cloud.githubusercontent.com/assets/10067146/7073136/914d4a4a-deae-11e4-88e4-2a4749d974f7.png)
![assignments_model_test](https://cloud.githubusercontent.com/assets/10067146/7073139/93b3b378-deae-11e4-8f3d-bfcec7938c99.png)

Hope this information helps whoever will pick up this issue.
